# IMPORT PANDAS

In [11]:
# Import Pandas

import pandas as pd

# Import actual_duration CSV As 'duration' Dataframe

In [45]:
# Import the actual_duration CSV File

duration = pd.read_csv('actual_duration.csv')

duration

sub_icb_location_code sub_icb_location_ons_code  \
0                        00L                 E38000130   
1                        00L                 E38000130   
2                        00L                 E38000130   
3                        00L                 E38000130   
4                        00L                 E38000130   
...                      ...                       ...   
137788                 X2C4Y                 E38000254   
137789                 X2C4Y                 E38000254   
137790                 X2C4Y                 E38000254   
137791                 X2C4Y                 E38000254   
137792                 X2C4Y                 E38000254   

                             sub_icb_location_name icb_ons_code  \
0       NHS North East and North Cumbria ICB - 00L    E54000050   
1       NHS North East and North Cumbria ICB - 00L    E54000050   
2       NHS North East and North Cumbria ICB - 00L    E54000050   
3       NHS North East and North Cumbria ICB - 00L    E54000050   
4       NHS North East and North Cumbria ICB - 00L    E54000050   
...                                            ...          ...   
137788              NHS West Yorkshire ICB - X2C4Y    E54000054   
137789              NHS West Yorkshire ICB - X2C4Y    E54000054   
137790              NHS West Yorkshire ICB - X2C4Y    E54000054   
137791              NHS West Yorkshire ICB - X2C4Y    E54000054   
137792              NHS West Yorkshire ICB - X2C4Y    E54000054   

       region_ons_code appointment_date         actual_duration  \
0            E40000012        01-Dec-21           31-60 Minutes   
1            E40000012        01-Dec-21           21-30 Minutes   
2            E40000012        01-Dec-21            6-10 Minutes   
3            E40000012        01-Dec-21  Unknown / Data Quality   
4            E40000012        01-Dec-21           16-20 Minutes   
...                ...              ...                     ...   
137788       E40000012        30-Jun-22           31-60 Minutes   
137789       E40000012        30-Jun-22           21-30 Minutes   
137790       E40000012        30-Jun-22           16-20 Minutes   
137791       E40000012        30-Jun-22           11-15 Minutes   
137792       E40000012        30-Jun-22             1-5 Minutes   

        count_of_appointments  
0                         364  
1                         619  
2                        1698  
3                        1277  
4                         730  
...                       ...  
137788                    430  
137789                    751  
137790                    921  
137791                   1439  
137792                   1739  

[137793 rows x 8 columns]

# Sense - Check 'duration' Dataframe

In [17]:
# Find the column names for the dataframe called 'duration'

print (duration.columns)

Index(['sub_icb_location_code', 'sub_icb_location_ons_code',
       'sub_icb_location_name', 'icb_ons_code', 'region_ons_code',
       'appointment_date', 'actual_duration', 'count_of_appointments'],
      dtype='object')


In [18]:
# Print the the dimensions for the dataframe called 'duration'

# The number of rows are printed first, followed by the number 
# of columns

print(duration.shape)

(137793, 8)


In [19]:
# Print the the data types contained within each column of
# the dataframe called 'duration'

print(duration.dtypes)

sub_icb_location_code        object
sub_icb_location_ons_code    object
sub_icb_location_name        object
icb_ons_code                 object
region_ons_code              object
appointment_date             object
actual_duration              object
count_of_appointments         int64
dtype: object


In [21]:
# Print the number of missing values within each column
# of the dataframe called 'duration'

print(duration.isnull().sum())

sub_icb_location_code        0
sub_icb_location_ons_code    0
sub_icb_location_name        0
icb_ons_code                 0
region_ons_code              0
appointment_date             0
actual_duration              0
count_of_appointments        0
dtype: int64


# Obtain Descriptive Statistics For 'duration' Dataframe

In [22]:
# Print the summary statistics for each column of
# the dataframe called 'duration'

duration.describe()

count_of_appointments
count          137793.000000
mean             1219.080011
std              1546.902956
min                 1.000000
25%               194.000000
50%               696.000000
75%              1621.000000
max             15400.000000

In [23]:
# Print the metadata of the dataframe called 'duration'

duration.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137793 entries, 0 to 137792
Data columns (total 8 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   sub_icb_location_code      137793 non-null  object
 1   sub_icb_location_ons_code  137793 non-null  object
 2   sub_icb_location_name      137793 non-null  object
 3   icb_ons_code               137793 non-null  object
 4   region_ons_code            137793 non-null  object
 5   appointment_date           137793 non-null  object
 6   actual_duration            137793 non-null  object
 7   count_of_appointments      137793 non-null  int64 
dtypes: int64(1), object(7)
memory usage: 8.4+ MB


# Import appointments_regional CSV As 'region' Dataframe

In [24]:
# Import the appointments_regional CSV File

region = pd.read_csv('appointments_regional.csv')

region

icb_ons_code appointment_month appointment_status hcp_type  \
0         E54000034           2020-01           Attended       GP   
1         E54000034           2020-01           Attended       GP   
2         E54000034           2020-01           Attended       GP   
3         E54000034           2020-01           Attended       GP   
4         E54000034           2020-01           Attended       GP   
...             ...               ...                ...      ...   
596816    E54000050           2022-06            Unknown  Unknown   
596817    E54000050           2022-06            Unknown  Unknown   
596818    E54000050           2022-06            Unknown  Unknown   
596819    E54000050           2022-06            Unknown  Unknown   
596820    E54000050           2022-06            Unknown  Unknown   

       appointment_mode time_between_book_and_appointment  \
0          Face-to-Face                             1 Day   
1          Face-to-Face                    15  to 21 Days   
2          Face-to-Face                       2 to 7 Days   
3          Face-to-Face                    22  to 28 Days   
4          Face-to-Face                     8  to 14 Days   
...                 ...                               ...   
596816          Unknown                       2 to 7 Days   
596817          Unknown                    22  to 28 Days   
596818          Unknown                     8  to 14 Days   
596819          Unknown                 More than 28 Days   
596820          Unknown                          Same Day   

        count_of_appointments  
0                        8107  
1                        6791  
2                       20686  
3                        4268  
4                       11971  
...                       ...  
596816                     21  
596817                      8  
596818                     28  
596819                     17  
596820                     10  

[596821 rows x 7 columns]

# Sense - Check 'region' Dataframe

In [25]:
# Find the column names for the dataframe called 'region'

print (region.columns)

Index(['icb_ons_code', 'appointment_month', 'appointment_status', 'hcp_type',
       'appointment_mode', 'time_between_book_and_appointment',
       'count_of_appointments'],
      dtype='object')


In [26]:
# Print the the dimensions for the dataframe called 'region'

# The number of rows are printed first, followed by the number 
# of columns

print(region.shape)

(596821, 7)


In [27]:
# Print the the data types contained within each column of
# the dataframe called 'region'

print(region.dtypes)

icb_ons_code                         object
appointment_month                    object
appointment_status                   object
hcp_type                             object
appointment_mode                     object
time_between_book_and_appointment    object
count_of_appointments                 int64
dtype: object


In [28]:
# Print the number of missing values within each column
# of the dataframe called 'region'

print(region.isnull().sum())

icb_ons_code                         0
appointment_month                    0
appointment_status                   0
hcp_type                             0
appointment_mode                     0
time_between_book_and_appointment    0
count_of_appointments                0
dtype: int64


# Obtain Descriptive Statistics For 'region' Dataframe

In [29]:
# Print the summary statistics for each column of
# the dataframe called 'region'

region.describe()

count_of_appointments
count          596821.000000
mean             1244.601857
std              5856.887042
min                 1.000000
25%                 7.000000
50%                47.000000
75%               308.000000
max            211265.000000

In [30]:
# Print the metadata of the dataframe called 'region'

region.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596821 entries, 0 to 596820
Data columns (total 7 columns):
 #   Column                             Non-Null Count   Dtype 
---  ------                             --------------   ----- 
 0   icb_ons_code                       596821 non-null  object
 1   appointment_month                  596821 non-null  object
 2   appointment_status                 596821 non-null  object
 3   hcp_type                           596821 non-null  object
 4   appointment_mode                   596821 non-null  object
 5   time_between_book_and_appointment  596821 non-null  object
 6   count_of_appointments              596821 non-null  int64 
dtypes: int64(1), object(6)
memory usage: 31.9+ MB


# Import national_categories excel As 'categories' Dataframe

In [51]:
# Import national_categories excel file

categories = pd.read_excel('national_categories.xlsx')

categories

appointment_date icb_ons_code  \
0            2021-08-02    E54000050   
1            2021-08-02    E54000050   
2            2021-08-02    E54000050   
3            2021-08-02    E54000050   
4            2021-08-02    E54000050   
...                 ...          ...   
817389       2022-06-30    E54000054   
817390       2022-06-30    E54000054   
817391       2022-06-30    E54000054   
817392       2022-06-30    E54000054   
817393       2022-06-30    E54000054   

                             sub_icb_location_name            service_setting  \
0       NHS North East and North Cumbria ICB - 00L       Primary Care Network   
1       NHS North East and North Cumbria ICB - 00L                      Other   
2       NHS North East and North Cumbria ICB - 00L           General Practice   
3       NHS North East and North Cumbria ICB - 00L           General Practice   
4       NHS North East and North Cumbria ICB - 00L           General Practice   
...                                            ...                        ...   
817389              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817390              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817391              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817392              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   
817393              NHS West Yorkshire ICB - X2C4Y  Extended Access Provision   

                  context_type                       national_category  \
0       Care Related Encounter  Patient contact during Care Home Round   
1       Care Related Encounter                         Planned Clinics   
2       Care Related Encounter                              Home Visit   
3       Care Related Encounter              General Consultation Acute   
4       Care Related Encounter            Structured Medication Review   
...                        ...                                     ...   
817389  Care Related Encounter             Unplanned Clinical Activity   
817390  Care Related Encounter                         Planned Clinics   
817391  Care Related Encounter              Planned Clinical Procedure   
817392  Care Related Encounter            General Consultation Routine   
817393  Care Related Encounter              General Consultation Acute   

        count_of_appointments appointment_month  
0                           3           2021-08  
1                           7           2021-08  
2                          79           2021-08  
3                         725           2021-08  
4                           2           2021-08  
...                       ...               ...  
817389                     12           2022-06  
817390                      4           2022-06  
817391                     92           2022-06  
817392                      4           2022-06  
817393                     19           2022-06  

[817394 rows x 8 columns]

# Sense - Check 'categories' Dataframe

In [38]:
# Find the column names for the dataframe called 'categories'

print (categories.columns)

Index(['appointment_date', 'icb_ons_code', 'sub_icb_location_name',
       'service_setting', 'context_type', 'national_category',
       'count_of_appointments', 'appointment_month'],
      dtype='object')


In [39]:
# Print the the dimensions for the dataframe called 'categories'

# The number of rows are printed first, followed by the number 
# of columns

print(categories.shape)

(817394, 8)


In [41]:
# Print the the data types contained within each column of
# the dataframe called 'categories'

print(categories.dtypes)

appointment_date         datetime64[ns]
icb_ons_code                     object
sub_icb_location_name            object
service_setting                  object
context_type                     object
national_category                object
count_of_appointments             int64
appointment_month                object
dtype: object


In [42]:
# Print the number of missing values within each column
# of the dataframe called 'categories'

print(categories.isnull().sum())

appointment_date         0
icb_ons_code             0
sub_icb_location_name    0
service_setting          0
context_type             0
national_category        0
count_of_appointments    0
appointment_month        0
dtype: int64


# Obtain Descriptive Statistics For 'categories' Dataframe

In [43]:
# Print the summary statistics for each column of
# the dataframe called 'categories'

categories.describe()

count_of_appointments
count          817394.000000
mean              362.183684
std              1084.576600
min                 1.000000
25%                 7.000000
50%                25.000000
75%               128.000000
max             16590.000000

In [44]:
# Print the metadata of the dataframe called 'categories'

categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817394 entries, 0 to 817393
Data columns (total 8 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   appointment_date       817394 non-null  datetime64[ns]
 1   icb_ons_code           817394 non-null  object        
 2   sub_icb_location_name  817394 non-null  object        
 3   service_setting        817394 non-null  object        
 4   context_type           817394 non-null  object        
 5   national_category      817394 non-null  object        
 6   count_of_appointments  817394 non-null  int64         
 7   appointment_month      817394 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 49.9+ MB


# Explore The Data

In [54]:
# Print the locations column from the 'categories' DataFrame

print("Count of locations: ", categories['sub_icb_location_name'])

Count of locations:  0         NHS North East and North Cumbria ICB - 00L
1         NHS North East and North Cumbria ICB - 00L
2         NHS North East and North Cumbria ICB - 00L
3         NHS North East and North Cumbria ICB - 00L
4         NHS North East and North Cumbria ICB - 00L
                             ...                    
817389                NHS West Yorkshire ICB - X2C4Y
817390                NHS West Yorkshire ICB - X2C4Y
817391                NHS West Yorkshire ICB - X2C4Y
817392                NHS West Yorkshire ICB - X2C4Y
817393                NHS West Yorkshire ICB - X2C4Y
Name: sub_icb_location_name, Length: 817394, dtype: object


In [82]:
# Print the distinct locations and their counts 


print("Count of locations: ", categories['sub_icb_location_name'].value_counts(ascending=True))

Count of locations:  NHS Greater Manchester ICB - 00V                2170
NHS Cheshire and Merseyside ICB - 01T           3242
NHS Cheshire and Merseyside ICB - 01V           3496
NHS Lancashire and South Cumbria ICB - 02G      4169
NHS North East and North Cumbria ICB - 00N      4210
                                               ...  
NHS North East London ICB - A3A8R              11837
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS Devon ICB - 15N                            12526
NHS Kent and Medway ICB - 91Q                  12637
NHS North West London ICB - W2U3Z              13007
Name: sub_icb_location_name, Length: 106, dtype: int64


In [83]:
# Print the distinct service settings and their counts 

print("Count of Service Settings: ", categories['service_setting'].value_counts(ascending=True))

Count of Service Settings:  Unmapped                      27419
Extended Access Provision    108122
Other                        138789
Primary Care Network         183790
General Practice             359274
Name: service_setting, dtype: int64


In [84]:
# Print the distinct context types and their counts 

print("Count of context types: ", categories['context_type'].value_counts(ascending=True))

Count of context types:  Unmapped                   27419
Inconsistent Mapping       89494
Care Related Encounter    700481
Name: context_type, dtype: int64


In [85]:
# Print the distinct national categories and their counts 

print("Count of national categories: ", categories['national_category'].value_counts(ascending=True))

Count of national categories:  Group Consultation and Group Education                                  5341
Walk-in                                                                14179
Non-contractual chargeable work                                        20896
Care Home Needs Assessment & Personalised Care and Support Planning    23505
Social Prescribing Service                                             26492
Care Home Visit                                                        26644
Unmapped                                                               27419
Patient contact during Care Home Round                                 28795
Unplanned Clinical Activity                                            40415
Home Visit                                                             41850
Service provided by organisation external to the practice              43095
Structured Medication Review                                           44467
Planned Clinical Procedure                   

In [86]:
# Print the distinct national categories and their counts 

print("Count of appointment statuses: ", region['appointment_status'].value_counts(ascending=True))

Count of appointment statuses:  DNA         163360
Unknown     201324
Attended    232137
Name: appointment_status, dtype: int64


# WEEK 2 QUESTIONS

#  1. 

How many locations are there in the data set?

106 distinct locations

817,394 locations in total, including duplicates

#  2.

What are the five locations with the highest number of records?

The 5 locations with the most records are:

                                    NHS North West London ICB - W2U3Z;
                                    NHS Kent and Medway ICB - 91Q;
                                    NHS Kent and Medway ICB - 91Q;  
                                    NHS Devon ICB - 15N;
                                    NHS Hampshire and Isle Of Wight ICB - D9Y0V  

#  3.

How many service settings, context types, national categories, and appointment statuses are there?

a. Number of Service Settings: 5 

b. Number of Context Types: 3 

c. Number of National Categories: 18              
    
d. Number of Appointment Statuses: 3 
    
    